runtime/server/controller.go

```diff
@@ -18,17 +24,106 @@ func (s *Server) WatchLogs(req *runtimev1.WatchLogsRequest, srv runtimev1.Runtim
 
 // ListResources implements runtimev1.RuntimeServiceServer
 func (s *Server) ListResources(ctx context.Context, req *runtimev1.ListResourcesRequest) (*runtimev1.ListResourcesResponse, error) {
-	panic("not implemented")
+	s.addInstanceRequestAttributes(ctx, req.InstanceId)
+	observability.AddRequestAttributes(ctx,
+		attribute.String("args.instance_id", req.InstanceId),
+		attribute.String("args.kind", req.Kind),
+	)
+
+	if !auth.GetClaims(ctx).CanInstance(req.InstanceId, auth.ReadObjects) {
+		return nil, ErrForbidden
+	}
+
+	ctrl, err := s.runtime.Controller(req.InstanceId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	rs, err := ctrl.List(ctx, req.Kind)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// TODO: Enforce security policy.
+
+	return &runtimev1.ListResourcesResponse{Resources: rs}, nil
 }
 
 // WatchResources implements runtimev1.RuntimeServiceServer
-func (s *Server) WatchResources(req *runtimev1.WatchResourcesRequest, srv runtimev1.RuntimeService_WatchResourcesServer) error {
-	panic("not implemented")
+func (s *Server) WatchResources(req *runtimev1.WatchResourcesRequest, ss runtimev1.RuntimeService_WatchResourcesServer) error {
+	observability.AddRequestAttributes(ss.Context(),
+		attribute.String("args.instance_id", req.InstanceId),
+		attribute.String("args.kind", req.Kind),
+	)
+
+	if !auth.GetClaims(ss.Context()).CanInstance(req.InstanceId, auth.ReadObjects) {
+		return ErrForbidden
+	}
+
+	ctrl, err := s.runtime.Controller(req.InstanceId)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	if req.Replay {
+		rs, err := ctrl.List(ss.Context(), req.Kind)
+		if err != nil {
+			return status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		// TODO: Enforce security policy.
+
+		for _, r := range rs {
+			err := ss.Send(&runtimev1.WatchResourcesResponse{
```

Comment:
We should add a lock around send.

Reply:
All calls to `ss.Send` happen in the same goroutine, so no lock is needed (added a comment on `Subscribe` to clarify this behavior)